contracts/OptimisticArbitrator.sol

```diff
@@ -0,0 +1,166 @@
+        request.proposedPrice = proposedPrice;
+        request.expirationTime = uint64(getCurrentTime()) + request.customLiveness;
+
+        request.currency.safeTransferFrom(msg.sender, address(this), request.bond + request.finalFee);
+    }
+
+    function disputePrice(uint256 timestamp, bytes memory ancillaryData) public {
+        Request storage request = requests[_getId(msg.sender, priceIdentifier, timestamp, ancillaryData)];
+        require(request.proposer != address(0), "No proposed price to dispute");
+        require(request.disputer == address(0), "Proposal already disputed");
+        require(uint64(getCurrentTime()) < request.expirationTime, "Proposal past liveness");
+
+        request.disputer = msg.sender;
+
+        request.currency.safeTransferFrom(msg.sender, address(this), request.bond + request.finalFee);
+
+        OptimisticOracleV2Interface oo = _getOptimisticOracle();
+
+        request.currency.approve(address(oo), 2 * (request.bond + request.finalFee) + request.reward);
+
+        oo.requestPrice(priceIdentifier, timestamp, ancillaryData, request.currency, request.reward);
+        oo.setBond(priceIdentifier, timestamp, ancillaryData, request.bond);
+        oo.proposePriceFor(
+            request.proposer,
+            address(this),
+            priceIdentifier,
+            timestamp,
+            ancillaryData,
+            request.proposedPrice
+        );
+        oo.disputePriceFor(msg.sender, address(this), priceIdentifier, timestamp, ancillaryData);
+
+        delete requests[_getId(msg.sender, priceIdentifier, timestamp, ancillaryData)];
+    }
+
+    function settleAndGetPrice(uint256 timestamp, bytes memory ancillaryData) public returns (int256) {
+        Request storage request = requests[_getId(msg.sender, priceIdentifier, timestamp, ancillaryData)];
+        require(address(request.currency) != address(0), "Price not requested");
+        require(request.proposer != address(0), "No proposed price to settle");
+        require(uint64(getCurrentTime()) > request.expirationTime, "Proposal not passed liveness");
+        require(request.disputer == address(0), "Proposal disputed, cant settle");
+
+        request.settled = true;
+
+        request.currency.safeTransfer(request.proposer, request.bond + request.finalFee + request.reward);
+
+        return request.proposedPrice;
+    }
+
+    function getPrice(uint256 timestamp, bytes memory ancillaryData) public view returns (int256) {
+        Request storage request = requests[_getId(msg.sender, priceIdentifier, timestamp, ancillaryData)];
+        require(request.settled == true, "Request not settled");
+        return request.proposedPrice;
+    }
+
+    function _getId(
+        address requester,
+        bytes32 identifier,
+        uint256 timestamp,
+        bytes memory ancillaryData
+    ) private pure returns (bytes32) {
+        return keccak256(abi.encodePacked(requester, identifier, timestamp, ancillaryData));
```

Comment:
nit: why do we need to consider identifier in this id/lookup if the identifier is always "YES_OR_NO_QUERY"?

Reply:
agreed. this can be simplified. I think we can just drip `identifier` from this and make the overall logic a bit cleaner.